base/src/com/google/idea/blaze/base/buildmodifier/BuildifierNotification.kt

```diff
@@ -0,0 +1,166 @@
+import com.intellij.notification.*
+import com.intellij.openapi.actionSystem.AnActionEvent
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.application.EDT
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.util.Key
+import com.intellij.platform.ide.progress.withBackgroundProgress
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+
+object BuildifierNotification {
+  private const val GITHUB_URL = "https://github.com/bazelbuild/buildtools/"
+
+  private const val NOTIFICATION_GROUP_ID = "Buildifier"
+  private const val NOTIFICATION_DOWNLOAD_TITLE = "Download Buildifier"
+  private const val NOTIFICATION_DOWNLOAD_QUESTION = "Would you like to download the buildifier formatter?"
+  private const val NOTIFICATION_DOWNLOAD_SUCCESS = "Buildifier download was successful."
+  private const val NOTIFICATION_DOWNLOAD_FAILURE = "Buildifier download failed. Please install it manually."
+  private const val NOTIFICATION_NOTFOUND_FAILURE = "Could not find buildifier binary. Please install it manually."
+
+  private const val PROGRESS_TITLE = "Downloading Buildifier binary..."
+
+  private val NOTIFICATION_DOWNLOAD_SHOWN_KEY = Key<Boolean>("buildifier.notification.download")
+  private val NOTIFICATION_NOTFOUND_SHOWN_KEY = Key<Boolean>("buildifier.notification.notfound")
+
+  private val NOTIFICATION_GROUP: NotificationGroup = NotificationGroupManager.getInstance().getNotificationGroup(
+    NOTIFICATION_GROUP_ID
+  )
+
+  private val OPEN_GITHUB_ACTION = NotificationAction.createSimple("Open GitHub Page") {
+    BrowserUtil.open(GITHUB_URL)
+  }
+
+  @JvmStatic
+  fun showDownloadNotification(project: Project) {
+    if (!shouldShowNotification(NOTIFICATION_DOWNLOAD_SHOWN_KEY)) {
+      return
+    }
+
+    val notification = NOTIFICATION_GROUP.createNotification(
+      NOTIFICATION_DOWNLOAD_TITLE,
+      NOTIFICATION_DOWNLOAD_QUESTION,
+      NotificationType.INFORMATION
+    ).setSuggestionType(true)
+
+    notification.addAction(
+      object : NotificationAction("Download") {
+        override fun actionPerformed(e: AnActionEvent, notification: Notification) {
+          notification.expire()
+          install(project)
+        }
+      }
+    )
+
+    notification.addDismissAction(NOTIFICATION_DOWNLOAD_SHOWN_KEY)
+
+    Notifications.Bus.notify(notification)
+  }
+
+  @JvmStatic
+  fun showNotFoundNotification() {
+    if (!shouldShowNotification(NOTIFICATION_NOTFOUND_SHOWN_KEY)) {
+      return
+    }
+
+    val notification = NOTIFICATION_GROUP.createNotification(
+      NOTIFICATION_NOTFOUND_FAILURE,
+      NotificationType.ERROR
+    )
+
+    notification.addAction(OPEN_GITHUB_ACTION)
+    notification.addDismissAction(NOTIFICATION_NOTFOUND_SHOWN_KEY)
+
+    Notifications.Bus.notify(notification)
+  }
+
+  private fun install(project: Project) {
+    val scope = pluginApplicationScope()
+
+    scope.launch {
```

Comment:
you can remove this scope and reuse `currentThreadCoroutineScope()` see https://youtrack.jetbrains.com/issue/IJPL-371 for the details